File: 56.合并区间.cpp
/*
 * @lc app=leetcode.cn id=56 lang=cpp
 *
 * [56] 合并区间
 */

#include <vector>
#include <algorithm>

class Solution {
public:
    std::vector<std::vector<int>> merge(std::vector<std::vector<int>>& intervals) {
        std::sort(intervals.begin(), intervals.end(), [](const auto& a, const auto& b){
            return a[0] < b[0];
        });
        std::vector<std::vector<int>> result;
        result.reserve(intervals.size());
        for(const auto& it : intervals){
            if(result.empty())
                result.emplace_back(it);
            else{
                if(result.back()[1] < it[0])
                    result.emplace_back(it);
                else
                    result.back()[1] = std::max(it[1], result.back()[1]);
            }
        }
        return result;
    }
};

File: 56.合并区间.rs
/*
 * @lc app=leetcode.cn id=56 lang=rust
 *
 * [56] 合并区间
 */

// @lc code=start
impl Solution {
    pub fn merge(mut intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        intervals.sort_unstable_by(|a, b| {
            return a[0].partial_cmp(&b[0]).unwrap();
        });
    
        let mut results = Vec::new();
        for s_e in intervals {
            if results.is_empty() {
                results.push(s_e);
            } else {
                let result = results.last_mut().unwrap();
                if result[1] >= s_e[0] {
                    result[1] = result[1].max(s_e[1]);
                } else {
                    results.push(s_e);
                }
            }
        }
    
        return results;
    }
}
// @lc code=end

